src/coprocessor/codec/convert.rs

```diff
@@ -287,26 +478,50 @@ fn float_str_to_int_string<'a, 'b: 'a>(
     }
 
     if e_idx.is_none() {
-        if int_cnt == 0 {
-            return Ok(Cow::Borrowed("0"));
+        let dot_idx = dot_idx.unwrap();
+        // NOTE: to make compatible with TiDB, +0.5 -> 1, -0.5 -> -1
+        let int_str = if int_cnt == 0 && valid_float.starts_with('-') {
+            &"-0"
+        } else if int_cnt == 0 {
+            &"0"
+        } else {
+            &valid_float[..dot_idx]
+        };
+        if digits_cnt - int_cnt > 0 {
+            // It's OK to unwrap here because the `dot_idx + 1` will less than the length of `valid_float`
+            let digit_char = valid_float.chars().nth(dot_idx + 1).unwrap();
+            return Ok(round_int_str(digit_char, int_str));
         }
-        return Ok(Cow::Borrowed(&valid_float[..dot_idx.unwrap()]));
+        return Ok(Cow::Borrowed(int_str));
     }
 
-    let exp = box_try!((&valid_float[e_idx.unwrap() + 1..]).parse::<i64>());
+    let e_idx = e_idx.unwrap();
+    let exp = box_try!((&valid_float[e_idx + 1..]).parse::<i64>());
     if exp > 0 && int_cnt > (i64::MAX - exp) {
-        // (exp + inc_cnt) overflows MaxInt64. Add warning and return original float string
+        // (exp + int_cnt) overflows MaxInt64. Add warning and return original float string
         ctx.warnings
             .append_warning(Error::overflow("BIGINT", &valid_float));
         return Ok(Cow::Owned(valid_float.to_owned()));
```

Comment:
Why this is not `Cow::Borrowed(valid_float)`

Reply:
Nice catch.